blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md

```diff
@@ -0,0 +1,317 @@
+```Haskell
+data MultiSet' = MultiSet' { _count :: Integer -> Integer }
+
+count' :: Integer -> MultiSet' -> Integer
+count' n m = _count m n
+
+empty :: MultiSet'
+empty = MultiSet' (\n -> 0)
+
+singleton :: Integer -> MultiSet'
+singleton n = MultiSet' (\m -> if n == m
+                               then 1
+                               else 0)
+
+union :: MultiSet' -> MultiSet' -> MultiSet'
+union s t = MultiSet' (\n -> (count' n s) + (count' n t))
+
+remove :: Integer -> MultiSet' -> MultiSet'
+remove n s = MultiSet' (\m -> if n == m
+                              then 0
+                              else count' n s)
+```
+
+And while we don't have decidable equality to help us this time, it's
+pretty obvious that `union` is commutative because `+` is commutative.
+Testing our example before works, no surprise:
+
+```Haskell
+> let s = singleton 1 `union` singleton 2
+> let t = singleton 2 `union` singleton 1
+> and [ count' n s == count' n t | n <- [0..1000]]
+True
+```
+
+How do we know this is the "best" or at least "most canonical"
+implementation of our datatype?
+The intuition is that we really don't care at all *how* our multisets
+are implemented as long as they behave the right way with respect to
+`count` since `count` returns an `Integer`, a type we do understand.
+This allows us to define "observational equivalence" of two multisets.
+
+We say that two closed terms `s,t` of sort `multiset` are
+*observationally equivalent* if for any term `C` of type `num` that
+has `s` as a subterm, we can swap `t` in for `s` and prove that the
+two terms are equal.
+For instance `C` might be `count(3, union(s, singleton(3)))` or
+`add(4,remove(5,s))`.
+Then we've reduced the possibly complicated equality for `multiset` to
+the simple equality of `num`.
+
+Proving that the final encoding above satisfies all observational
+equivalences is beyond the scope of this blog post (see
+[here][proving-final-encoding]), but let's see what all this talk
+about "algebras", initial or final is all about.
+
+# Formalization Attempt 1: Algebras of a Theory
+
+First, our little language of numbers and multisets is called a
+*theory*. The specific category gadget that we'll use to describe it
+is a *multi-sorted Lawvere theory*, or just *Lawvere theory* for
+short.
+
+A *Lawvere theory* is a category with finite products all of whose
+objects are finite products of a collection of *sorts* \\(S\\). We can
+construct this category from our little language above by making the
+objects be *contexts* \\(x:num,y:multiset,...\\) and morphisms \\(\Gamma ->
+x_1:s_1,...,x_n:s_n\\) to be \\(n\\)-tuples of terms \\(\Gamma -> s_1,...,
+\Gamma -> s_n\\) *modulo* the equations we've specified. We'll use the
+letter \\(T\\) to mean a Lawvere theory.
+
+Then a *\\(T\\)-algebra* is a denotational semantics of our theory \\(T\\),
+i.e., a product preserving functor \\(A : T \to Set\\). This means for
+every sort we get a set \\(A(s)\\) and for every term \\(x_1:s_1,...,x_n:s_n
```

Comment:
hmm so "sorts" and "terms" are 2 kinds of objects in `T`?

Reply:
sorts are the objects, terms are the arrows